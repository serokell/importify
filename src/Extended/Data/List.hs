-- | This module contains additional utility functions for list.

module Extended.Data.List
       ( concatNub
       , foldMaybeMap
       , isIn
       , removeAt
       , removeAtMultiple
       ) where

import           Universum

-- | Removes element from list by given index. If there's
-- no element at such index then list returns unchanged.
removeAt :: Int -> [a] -> [a]
removeAt i l | i < 0 = l
removeAt i l = case after of
                     []       -> l
                     (_:rest) -> before ++ rest
  where (before, after) = splitAt i l

-- | Like 'removeAt' but takes list of indices to remove.
removeAtMultiple :: [Int] -> [a] -> [a]
removeAtMultiple indices = map snd
                         . filter ((`notElem` indices) . fst)
                         . zip [0..]

-- | Folds list of values with Maybe monoidal function.
foldMaybeMap :: Monoid m => (a -> Maybe m) -> [a] -> m
foldMaybeMap f = foldMap (maybeToMonoid . f)

-- | Flipped version of 'elem'.
isIn :: Eq a => [a] -> a -> Bool
isIn = flip elem
{-# INLINE isIn #-}

-- | Like 'concatMap' but also applies 'ordNub' to resulted list.
concatNub :: Ord b => (a -> [b]) -> [a] -> [b]
concatNub = ordNub ... concatMap
