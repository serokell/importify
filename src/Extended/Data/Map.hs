{-# LANGUAGE ExplicitForAll      #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeFamilies        #-}

module Extended.Data.Map
       ( groupBy
       , insertNonEmpty
       , invertMapToList
       , mapBy
       ) where

import Universum

import qualified Data.Map as Map
import qualified Data.Semigroup as Semigroup ((<>))

-- | Inserts one element to map from keys to 'NonEmpty' list.
insertNonEmpty :: Ord k => k -> v -> Map k (NonEmpty v) -> Map k (NonEmpty v)
insertNonEmpty k v = Map.insertWith (flip (Semigroup.<>)) k (one v)

-- | Groups 'NonTrivialContainer' into 'Map' by given function.
groupBy :: forall f a b .
           (NontrivialContainer f, Element f ~ a, Eq b, Ord b)
        => (a -> b)
        -> f
        -> Map b (NonEmpty a)
groupBy f = foldl' insertBy mempty
  where
    insertBy :: Map b (NonEmpty a) -> a -> Map b (NonEmpty a)
    insertBy dict value = insertNonEmpty (f value) value dict

-- | Like 'groupBy' but keeps only one element of mapping.
mapBy :: forall f a b .
         (NontrivialContainer f, Element f ~ a, Eq b, Ord b)
      => (a -> b)
      -> f
      -> Map b a
mapBy f = foldl' insertBy mempty
  where
    insertBy :: Map b a -> a -> Map b a
    insertBy dict value = Map.insert (f value) value dict

-- | Creates Map from values to keys by mapping each value to corresponding key.
invertMapToList :: forall k v . (Ord k, Ord v) => Map k [v] -> Map v k
invertMapToList = foldl' insertList mempty . Map.toList
  where
    insertList :: Map v k -> (k, [v]) -> Map v k
    insertList valueMap (key, values) =
        foldl' (\vals val -> Map.insert val key vals) valueMap values
