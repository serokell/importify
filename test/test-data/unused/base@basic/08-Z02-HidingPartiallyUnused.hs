module HidingPartiallyUnused where

import           Data.List   hiding (sort, unionBy)
import           GHC.OldList (sort)

foo = sort
bar = subsequences
