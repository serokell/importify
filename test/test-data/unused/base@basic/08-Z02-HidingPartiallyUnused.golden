module HidingPartiallyUnused where

import           Data.List   hiding (sort         )
import           GHC.OldList (sort)

foo = sort
bar = subsequences
